File: GoLang/solution/solution0208.go
package solution

type Trie struct {
	child []*Trie
	isEnd bool
}

func Constructor0208() Trie {
	return Trie{make([]*Trie, 26), false}
}

func (this *Trie) Insert(word string) {
	node := this
	for _, value := range word {
		if node.child[value-'a'] == nil {
			node.child[value-'a'] = &Trie{make([]*Trie, 26), false}
		}
		node = node.child[value-'a']
	}
	node.isEnd = true
}

func (this *Trie) Search(word string) bool {
	node := this
	for _, value := range word {
		node = node.child[value-'a']
		if node == nil {
			return false
		}
	}
	return node.isEnd
}

func (this *Trie) StartsWith(prefix string) bool {
	node := this
	for _, value := range prefix {
		node = node.child[value-'a']
		if node == nil {
			return false
		}
	}
	return true
}

/**
 * Your Trie object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(word);
 * param_2 := obj.Search(word);
 * param_3 := obj.StartsWith(prefix);
 */

File: JavaR3/src/solution/Solution0208.java
package solution;

public class Solution0208 {
	class Trie {

		boolean isEnd;
		Trie[] next;

		public Trie() {
			isEnd = false;
			next = new Trie[26];
		}

		public void insert(String word) {
			Trie node = this;
			for (int i = 0; i < word.length(); i++) {
				char c = word.charAt(i);
				if (node.next[c - 'a'] == null) {
					node.next[c - 'a'] = new Trie();
				}

				node = node.next[c - 'a'];
			}

			node.isEnd = true;
		}

		public boolean search(String word) {
			Trie node = this;

			for (int i = 0; i < word.length(); i++) {
				char c = word.charAt(i);
				if (node.next[c - 'a'] == null) {
					return false;
				}
				node = node.next[c - 'a'];
			}

			return node.isEnd;
		}

		public boolean startsWith(String prefix) {
			Trie node = this;

			for (int i = 0; i < prefix.length(); i++) {
				char c = prefix.charAt(i);
				if (node.next[c-'a'] == null) {
					return false;
				}
				node = node.next[c - 'a'];
			}

			return true;
		}
	}

}
